samples/streaming/provider/src/provider_impl.rs

```diff
@@ -0,0 +1,214 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+// SPDX-License-Identifier: MIT
+
+// use digital_twin_model::sdv_v1 as sdv;
+use core::iter::Iterator;
+use log::{debug, info, warn};
+use samples_protobuf_data_access::sample_grpc::v1::digital_twin_provider::digital_twin_provider_server::DigitalTwinProvider;
+use samples_protobuf_data_access::sample_grpc::v1::digital_twin_provider::{
+    GetRequest, GetResponse, InvokeRequest, InvokeResponse, SetRequest, SetResponse,
+    SubscribeRequest, SubscribeResponse, UnsubscribeRequest, UnsubscribeResponse,
+    StreamRequest, StreamResponse};
+use std::fs::File;
+use std::io::{Error, Read};
+use std::path::Path;
+use std::pin::Pin;
+use std::vec::Vec;
+use tokio::sync::mpsc;
+use tokio_stream::{wrappers::ReceiverStream, Stream, StreamExt};
+use tokio::time::Duration;
+use tonic::{Request, Response, Status};
+use uuencode::uuencode;
+
+#[derive(Debug, Default)]
+/// Continuously iterates through a list of files.
+/// After it reaches the end of the list, it cycles back to the start of the list.
+struct ImageFileIterator {
+    /// The directory that contains the images.
+    image_directory: String,
+    /// The filenames of the images that we want to stream.
+    image_filenames: Vec<String>,
+    /// The index for the current image filename.
+    current_image_file_index: usize,
+}
+
+impl ImageFileIterator {
+    /// Create a new ImageFileIterator.
+    ///
+    /// # Arguments
+    /// * `image_directory` - The directory that contains the images.
+    /// * `image_filenames` - The filenames of the images that we want to stream.
+    pub fn new(image_directory: &str, image_filenames: Vec<String>) -> Self {
+        Self {
+            image_directory: image_directory.to_string(),
+            image_filenames,
+            current_image_file_index: 0,
+        }
+    }
+
+    /// Reads an image file and returns its uuencoded form.
+    ///
+    /// # Arguments
+    /// * `filenme` -The image's filename.
+    fn read_image_file(&self, filename: &str) -> Result<String, Error> {
+        let filepath = Path::new(&self.image_directory).join(filename);
+        debug!("Read_image from '{}'", filepath.display());
+        let mut file = File::open(filepath)?;
+        let mut file_content = Vec::new();
+        file.read_to_end(&mut file_content).expect("Unable to read");
+        let encoded = uuencode(filename, &file_content);
```

Comment:
Why this type of encoding over base64 or byte array?

Reply:
The sample now transfers bytes and a media/mime type, like Chariott's dog mode demo.